resolver/resolver-tcp.go

```diff
@@ -0,0 +1,359 @@
+	}
+	return tr
+}
+
+func (tr *TCPResolver) client(workerCtx context.Context) error { //nolint:gocognit,gocyclo // TODO
+	connTimer := time.NewTimer(tr.clientTTL)
+	connClosing := abool.New()
+	var connCtx context.Context
+	var cancelConnCtx func()
+	var recycleConn bool
+	var shuttingDown bool
+	var incoming = make(chan *dns.Msg, 100)
+
+	// enable client restarting after crash
+	defer tr.clientStarted.UnSet()
+
+connMgmt:
+	for {
+		// cleanup old connection
+		if tr.dnsConnection != nil {
+			connClosing.Set()
+			_ = tr.dnsConnection.Close()
+			cancelConnCtx()
+
+			tr.dnsConnection = nil
+			atomic.AddUint32(tr.connInstanceID, 1)
+		}
+
+		// check if we are shutting down or failing
+		if shuttingDown || tr.IsFailing() {
+			// reply to all waiting queries
+			tr.Lock()
+			for id, inFlight := range tr.inFlightQueries {
+				close(inFlight.Response)
+				delete(tr.inFlightQueries, id)
+			}
+			tr.clientStarted.UnSet() // in lock to guarantee to set before submitQuery proceeds
+			tr.Unlock()
+
+			// hint network environment at failed connection
+			netenv.ReportFailedConnection()
+
+			cancelConnCtx() // The linter said so. Don't even...
+			return nil
+		}
+
+		// wait until there is something to do
+		tr.Lock()
+		waiting := len(tr.inFlightQueries)
+		tr.Unlock()
+		if waiting > 0 {
+			// queue abandoned queries
+			ignoreBefore := time.Now().Add(-ignoreQueriesAfter)
+			currentConnInstanceID := atomic.LoadUint32(tr.connInstanceID)
+			tr.Lock()
+			for id, inFlight := range tr.inFlightQueries {
+				if inFlight.Started.Before(ignoreBefore) {
+					// remove
+					delete(tr.inFlightQueries, id)
+				} else if inFlight.ConnInstanceID != currentConnInstanceID {
+					inFlight.ConnInstanceID = currentConnInstanceID
+					// re-inject
+					select {
+					case tr.queries <- inFlight.Msg:
+					default:
+						log.Warningf("resolver: failed to re-inject abandoned query to %s", tr.resolver.Name)
+					}
+				}
+			}
+			tr.Unlock()
+		} else {
+			// wait for first query
+			select {
+			case <-workerCtx.Done():
+				// abort
+				shuttingDown = true
+				continue connMgmt
+			case msg := <-tr.queries:
+				// re-insert, we will handle later
+				select {
+				case tr.queries <- msg:
+				default:
+					log.Warningf("resolver: failed to re-inject waking query to %s", tr.resolver.Name)
+				}
+			}
+		}
+
+		// create connection
+		connCtx, cancelConnCtx = context.WithCancel(workerCtx)
+		// refresh dialer for authenticated local address
+		tr.dnsClient.Dialer = &net.Dialer{
+			LocalAddr: getLocalAddr("tcp"),
+			Timeout:   defaultConnectTimeout,
+			KeepAlive: defaultClientTTL,
+		}
+		// connect
+		c, err := tr.dnsClient.Dial(tr.resolver.ServerAddress)
+		if err != nil {
+			tr.ReportFailure()
+			log.Debugf("resolver: failed to connect to %s (%s)", tr.resolver.Name, tr.resolver.ServerAddress)
+			continue connMgmt
+		}
+		tr.dnsConnection = c
+		log.Debugf("resolver: connected to %s (%s)", tr.resolver.Name, tr.dnsConnection.RemoteAddr())
```

Comment:
no locking required when setting those values on `tr`?  Seems like a good candidate for a dedicated method.